### DISCUSSION TITLE
How to import zeromq from node_modules?

### DISCUSSION THREAD

**Discussion by AfdhalAtiffTan** [pocketbase/pocketbase#4690] [Q&A]
Hello,

Is it possible to use a module installed via NPM in a PocketBase's hook?

I am not sure where to run `npm i zeromq` and how to import it in a hook.

Can I specify a path for the `node_modules` folder manually?

Thank you.

**Comment by ganigeorgiev** [comment]
The JS `pb_hooks` currently supports loading CJS modules but only if they don't rely on runtime specific APIs (see [Loading modules](https://pocketbase.io/docs/js-overview/#loading-modules)).

From a quick look at `zeromq` I don't think it will work with PocketBase as it probably needs some low level Node.js API to interact with the zeromq library.

**Reply by AfdhalAtiffTan** [reply]
Hi Gani, 

Thank you for your reply and the link.

In case I want to try it, just to confirm, do I run `npm install` commands in the `pb_hooks` folder so that the `node_modules` folder is inside it?

**Reply by bartonhammond** [reply]
Did you ever figure this out?

**Reply by AfdhalAtiffTan** [reply]
> Did you ever figure this out?

Unfortunately not, I use a simple nodejs+express endpoint as a workaround.

**Reply by bartonhammond** [reply]
I did earlier today. - see [https://github.com/pocketbase/pocketbase/discussions/6215](https://github.com/pocketbase/pocketbase/discussions/6215)

In case you're interested, here's a repo that I'm going to use w/ the HTMX tutorial - it is using JS in the Extend style - no Express or using the API - just using PB as a framework.  I think it's my new stack.  I haven't done anything yet w/ the HTMX part - the repo is at the starting point for the tutorial from here: [https://hypermedia.systems/book/contents/](https://hypermedia.systems/book/contents/)

repo is here: [https://github.com/bartonhammond/htmx-training](https://github.com/bartonhammond/htmx-training) - See [https://github.com/bartonhammond/htmx-training/blob/master/pb_hooks/new.pb.js](https://github.com/bartonhammond/htmx-training/blob/master/pb_hooks/new.pb.js)

**Reply by ganigeorgiev** [reply]
@bartonhammond You are not using zeromq.

I'm almost certain that `zeromq` cannot be loaded with the PocketBase JSVM because it relies on Node.js primitives (filesystem, sockets, C interop, etc.), aka. it is not a pure ES5 module.

The problem in the discussion that you've linked is explained in https://pocketbase.io/docs/js-overview/#handlers-scope.

_As a side-note: I don't recommend using PocketBase with HTMX as you might find yourself implementing a lot of utilities on your own, which kindof defeat the purpose of PocketBase in the first place. Security consideration is also another reason that many users in the past fails to implement properly and once again we currently don't have the necessary builtin helpers for SSR apps. If you haven't already, you can also check the recommendation in https://pocketbase.io/docs/client-side-sdks/._

**Reply by bartonhammond** [reply]
my bad - misunderstood
